### ISSUE TITLE
core v.0.12.1 Failed install

### ISSUE THREAD

**Issue by gabefighter99** [janestreet/core#127]
MacBook Air
macOS High Sierra v10.13.6

`$ uname -a`

Output:
`Darwin YNC28012019P11.local 17.7.0 Darwin Kernel Version 17.7.0: Thu Dec 20 21:47:19 PST 2018; root:xnu-4570.71.22~1/RELEASE_X86_64 x86_64`

Reproducible steps:
```
$ git clone https://github.com/janestreet/core.git

$ opam install .

core is now pinned to git+file:///Users/cipe/Desktop/gitfiles/core#master (version v0.12.0)
The following actions will be performed:
  - recompile core v0.12.0*

<><> Gathering sources ><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
[core.v0.12.0] synchronised from git+file:///Users/cipe/Desktop/gitfiles/core#master

<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
[ERROR] The compilation of core failed at
        "/Users/cipe/.opam/opam-init/hooks/sandbox.sh build dune build -p core
        -j 3".

#=== ERROR while compiling core.v0.12.0 =======================================#
# context     2.0.3 | macos/x86_64 | ocaml-system.4.07.1 | pinned(git+file:///Users/cipe/Desktop/gitfiles/core#master#a173718c)
# path        ~/.opam/default/.opam-switch/build/core.v0.12.0
# command     ~/.opam/opam-init/hooks/sandbox.sh build dune build -p core -j 3
# exit-code   1
# env-file    ~/.opam/log/core-2612-d0f245.env
# output-file ~/.opam/log/core-2612-d0f245.out
### output ###
# File "src/dune", line 7, characters 2-19:
# 7 |   core_kernel.flags sexplib.unix spawn threads core_kernel.version_util)
#       ^^^^^^^^^^^^^^^^^
# Error: Library "core_kernel.flags" not found.
# Hint: try: dune external-lib-deps --missing -j 3 -p core @@default



<><> Error report <><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
+- The following actions failed
| - build core v0.12.0
+- 
- No changes have been performed

$ dune external-lib-deps --missing -j 3 -p core @@default
Error: The following libraries are missing in the default context:
- core_kernel.flags
- core_kernel.version_util
Hint: try: opam install core_kernel

$ opam install core_kernel
[NOTE] Package core_kernel is already installed (current version is v0.12.0).

$ dune external-lib-deps --missing -j 3 -p core @@default
Error: The following libraries are missing in the default context:
- core_kernel.flags
- core_kernel.version_util
Hint: try: opam install core_kernel
```

Not sure how to proceed...

**Comment by diml** [comment]
What version of core_kernel do you have installed?

**Comment by gabefighter99** [comment]
`[NOTE] Package core_kernel is already installed (current version is v0.12.0).`

Isn't that it?

**Comment by diml** [comment]
Are you trying to build the master of core against core_kernel.v0.12.0? This is not supported. If you want to do that, you need to add our bleeding edge opam repository:

```
$ opam repo add js-bleeding https://github.com/janestreet/opam-repository
```

**Comment by gabefighter99** [comment]
And otherwise? Is there an updated version of core_kernel that I can't find? I thought mine was the latest

**Comment by diml** [comment]
If you want to build the core#master, then you need core_kernel#master. That's what janestreet/opam-repository provides.

**Comment by gabefighter99** [comment]
Update:

`$ opam repo add js-bleeding https://github.com/janestreet/opam-repository`

Ran the above code, failed with the following:
```
[ERROR] Could not update repository "js-bleeding":
        OpamDownload.Download_fail(_, "curl: code 404 while downloading
        https://github.com/janestreet/opam-repository/index.tar.gz")
[ERROR] Initial repository fetch failed
```

Then ran:
`$ opam repo add janestreet-bleeding https://ocaml.janestreet.com/opam-repository`

The above seemed to work fine.

So then I ran:
`$ opam update`
`$ opam upgrade`

Both of which, succeeded, but the latter couldn't install core, due to something familiar:
```
[ERROR] The compilation of core failed at
        "/Users/cipe/.opam/opam-init/hooks/sandbox.sh build dune build -p core
        -j 3".

#=== ERROR while compiling core.v0.12.0 =======================================#
# context     2.0.3 | macos/x86_64 | ocaml-system.4.07.1 | pinned(git+file:///Users/cipe/Desktop/gitfiles/core#master#a173718c)
# path        ~/.opam/default/.opam-switch/build/core.v0.12.0
# command     ~/.opam/opam-init/hooks/sandbox.sh build dune build -p core -j 3
# exit-code   1
# env-file    ~/.opam/log/core-7616-83eed9.env
# output-file ~/.opam/log/core-7616-83eed9.out
### output ###
# File "src/dune", line 7, characters 2-19:
# 7 |   core_kernel.flags sexplib.unix spawn threads core_kernel.version_util)
#       ^^^^^^^^^^^^^^^^^
# Error: Library "core_kernel.flags" not found.
# Hint: try: dune external-lib-deps --missing -j 3 -p core @@default
```

As I've done a loop, I'm once again at a loss. "https://github.com/janestreet/opam-repository/index.tar.gz" doesn't exist. I think core has been removed from opam for the time-being.

from opam update:
```
+- The following changes have been performed
| - remove    core                    v0.12.0
| - ...
```

And seriously, thank-you for how helpful you've been.

**Comment by gabefighter99** [comment]
Showing output of:
`$ opam info core_kernel`

Output:
```
<><> core_kernel: information on all versions <><><><><><><><><><><><><><><><><>
name                   core_kernel
all-installed-versions v0.12.0 [default]
all-versions           109.27.00  109.28.00  109.30.00  109.31.00  109.32.00
                       109.33.00  109.34.00  109.35.00  109.35.01  109.36.00
                       109.37.00  109.38.00  109.40.00  109.41.00  109.42.00
                       109.45.00  109.47.00  109.53.00  109.55.00  109.55.02
                       109.58.00  109.60.00  110.01.00  111.03.00  111.06.00
                       111.08.00  111.11.00  111.13.00  111.17.00  111.21.00
                       111.25.00  111.28.00  112.01.00  112.06.00  112.06.02
                       112.17.00  112.24.00  112.35.00  113.00.00  113.24.00
                       113.33.00  113.33.01  113.33.01+4.03  113.33.02+4.03
                       113.33.03  v0.9.0  v0.9.1  v0.10.0  v0.11.0  v0.11.1
                       v0.12.0  v0.13-preview.120.28+03

<><> Version-specific details <><><><><><><><><><><><><><><><><><><><><><><><><>
version       v0.12.0
repository    default
url.src:      "https://ocaml.janestreet.com/ocaml-core/v0.12/files/core_kernel-v0.12.0.tar.gz"
url.checksum: "md5=e7adb97d39a93ccd542b9d015659c480"
homepage:     "https://github.com/janestreet/core_kernel"
bug-reports:  "https://github.com/janestreet/core_kernel/issues"
dev-repo:     "git+https://github.com/janestreet/core_kernel.git"
authors:      "Jane Street Group, LLC <opensource@janestreet.com>"
maintainer:   "opensource@janestreet.com"
license:      "MIT"
depends:      "ocaml" {>= "4.07.0"}
              "base" {>= "v0.12" & < "v0.13"}
              "base_bigstring" {>= "v0.12" & < "v0.13"}
              "base_quickcheck" {>= "v0.12" & < "v0.13"}
              "bin_prot" {>= "v0.12" & < "v0.13"}
              "jst-config" {>= "v0.12" & < "v0.13"}
              "fieldslib" {>= "v0.12" & < "v0.13"}
              "jane-street-headers" {>= "v0.12" & < "v0.13"}
              "ppx_assert" {>= "v0.12" & < "v0.13"}
              "ppx_base" {>= "v0.12" & < "v0.13"}
              "ppx_hash" {>= "v0.12" & < "v0.13"}
              "ppx_inline_test" {>= "v0.12" & < "v0.13"}
              "ppx_jane" {>= "v0.12" & < "v0.13"}
              "ppx_sexp_conv" {>= "v0.12" & < "v0.13"}
              "ppx_sexp_message" {>= "v0.12" & < "v0.13"}
              "sexplib" {>= "v0.12" & < "v0.13"}
              "splittable_random" {>= "v0.12" & < "v0.13"}
              "stdio" {>= "v0.12" & < "v0.13"}
              "time_now" {>= "v0.12" & < "v0.13"}
              "typerep" {>= "v0.12" & < "v0.13"}
              "variantslib" {>= "v0.12" & < "v0.13"}
              "dune" {build & >= "1.5.1"}
synopsis      Industrial strength alternative to OCaml's standard library
description
          The Core suite of libraries is an industrial strength alternative to
          OCaml's standard library that was developed by Jane Street, the
          largest industrial user of OCaml.
          Core_kernel is the system-independent part of Core.
```

**Comment by diml** [comment]
Yh, the error message is confusing, however the output of `opam info core_kernel` does indicate that it is v0.12.0 that is installed. I guess it is because we don't generate constraint in the opam file that is core's repository.

**Comment by diml** [comment]
So `opam pin` thing it can use core.v0.12.0...

**Comment by gabefighter99** [comment]
So how should I proceed in that case?

**Comment by gabefighter99** [comment]
IT WORKS!!

So I couldn't for the life of me figure out why core_kernel simultaneously was and wasn't there, when running, as suggested, `opam install core_kernel`, opam kept telling me that I already had it installed.

It turns out, I think, whenever I was running `opam install core`, it would use a pinned \core library on my computer, since initially I tried downloading from there. I just went to that folders root and did:

`opam uninstall .`

Deleted the file and ran `opam install core`. Core installed v0.12.1, something that hadn't happened before and now everything's running.

Thanks for your help!
